Allow Tensor& in the unboxing logic

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42841 Remove impl_unboxedOnlyKernel
* **#42712 Allow Tensor& in the unboxing logic**

Previously, operators taking Tensor& as arguments or returning it couldn't be c10-full because the unboxing logic didn't support it.
This adds temporary support for that. We're planning to remove this again later, but for now we need it to make those ops c10-full.
See https://docs.google.com/document/d/19thMVO10yMZA_dQRoB7H9nTPw_ldLjUADGjpvDmH0TQ for the full plan.

This PR also makes some ops c10-full that now can be.

Differential Revision: [D22989242](https://our.internmc.facebook.com/intern/diff/D22989242/)

smessmer committed:
[wip] Allow Tensor& in the unboxing logic

Previously, operators taking Tensor& as arguments or returning it couldn't be c10-full because the unboxing logic didn't support it.
This adds temporary support for that. We're planning to remove this again later, but for now we need it to make those ops c10-full.
See https://docs.google.com/document/d/19thMVO10yMZA_dQRoB7H9nTPw_ldLjUADGjpvDmH0TQ for the full plan.

This PR also makes some ops c10-full that now can be.

Differential Revision: [D22989242](https://our.internmc.facebook.com/intern/diff/D22989242/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] Allow Tensor& in the unboxing logic"


Previously, operators taking Tensor& as arguments or returning it couldn't be c10-full because the unboxing logic didn't support it.
This adds temporary support for that. We're planning to remove this again later, but for now we need it to make those ops c10-full.
See https://docs.google.com/document/d/19thMVO10yMZA_dQRoB7H9nTPw_ldLjUADGjpvDmH0TQ for the full plan.

This PR also makes some ops c10-full that now can be.

Differential Revision: [D22989242](https://our.internmc.facebook.com/intern/diff/D22989242/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] Allow Tensor& in the unboxing logic"


Previously, operators taking Tensor& as arguments or returning it couldn't be c10-full because the unboxing logic didn't support it.
This adds temporary support for that. We're planning to remove this again later, but for now we need it to make those ops c10-full.
See https://docs.google.com/document/d/19thMVO10yMZA_dQRoB7H9nTPw_ldLjUADGjpvDmH0TQ for the full plan.

This PR also makes some ops c10-full that now can be.

Differential Revision: [D22989242](https://our.internmc.facebook.com/intern/diff/D22989242/)

[ghstack-poisoned]

bhosmer commented on aten/src/ATen/core/boxing/impl/make_boxed_from_unboxed_functor.h:
```diff
@@ -307,9 +335,10 @@ using supported_primitive_arg_types = guts::typelist::typelist<
       guts::if_constexpr<has_outputs>([&] (auto _) {
-        auto output = call_functor_with_args_from_stack<KernelFunctor, AllowDeprecatedTypes>(functor_, _(stack));
+        using ReturnType_ = std::remove_cv_t<std::remove_reference_t<typename decltype(_)::template type_identity<ReturnType>>>;
+        ReturnType_ output = call_functor_with_args_from_stack<KernelFunctor, AllowDeprecatedTypes>(functor_, _(stack));
```
Also personally I still think using `_` for the typecheck-delaying lambda makes this code even tougher to grok if you're not pretty deep into templates already. I wouldn't push on it if you like the look, but worth noting its consequences in terms of audience :)

smessmer replied:
renamed `_` to `delay_check`.

bhosmer commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -1151,7 +1139,7 @@ IValue from_(T x, std::false_type) {
 template <typename T>
 IValue from(T x) {
-  return detail::from_(std::move(x), detail::has_constructor<T>{});
+  return detail::from_(std::move(x), typename std::is_constructible<IValue, T>::type{});
```
Is it newly possible to use `is_constructible` instead of `has_constructor`, or is this just cleanup?

smessmer replied:
the old implementation ran into compiler bugs on nvcc but is_constructible would always have worked, so it's just a cleanup.

bhosmer reviewed (approved):
Cool, this is a big step! 90% of the inline stuff is formatting/commenting requests, all at your discretion. And the rest is just me wanting to confirm my understanding of why the approach works :)

smessmer commented on aten/src/ATen/core/boxing/impl/make_boxed_from_unboxed_functor.h:
```diff
@@ -254,28 +254,56 @@ using supported_primitive_arg_types = guts::typelist::typelist<
   struct push_outputs final {
     static void call(OutputType&& output, Stack* stack) {
-      torch::jit::push(*stack, return_to_ivalue<OutputType, AllowDeprecatedTypes>(std::move(output)));
+      torch::jit::push(*stack, return_to_ivalue<OutputType, AllowDeprecatedTypes>(std::forward<OutputType>(output)));
```
I think this was required because it could now be a `Tensor&` which can't be std::move'd.

smessmer commented on aten/src/ATen/core/boxing/impl/make_boxed_from_unboxed_functor.h:
```diff
@@ -254,28 +254,56 @@ using supported_primitive_arg_types = guts::typelist::typelist<
+     * and C++ doesn't allow us to call (*functor) with a temporary "Tensor" when it expects "Tensor&".
+     * We use reference_cast to explicitly cast our temporary to a "Tensor&" and make it pass the compiler.
+     * Even though usually dangerous, this is ok here because temporaries live until the end of the statement.
+     * TODO We should remove reference_cast once kernels don't take "Tensor&" arguments anymore
```
that's what the `remove_reference_t` in the return type is for. It makes sure we cast any potential `Tensor&` to an actual `Tensor` value.

smessmer commented on aten/src/ATen/core/boxing/impl/make_boxed_from_unboxed_functor.h:
```diff
@@ -254,28 +254,56 @@ using supported_primitive_arg_types = guts::typelist::typelist<
+    return (*functor)(reference_cast<guts::typelist::element_t<ivalue_arg_indices, ArgTypes>>(
+      ivalue_to_arg<std::remove_cv_t<std::remove_reference_t<guts::typelist::element_t<ivalue_arg_indices, ArgTypes>>>, AllowDeprecatedTypes>::call(
+        std::move(torch::jit::peek(*stack, ivalue_arg_indices, num_ivalue_args))
+    ))...);
```
`std::remove_cv_t<std::remove_reference_t<T>>` is mostly equivalent to `std::decay_t<T>`. For the scope of our use cases, they should be equivalent. I'll change it to `decay_t`.

smessmer commented on aten/src/ATen/core/boxing/impl/make_boxed_from_unboxed_functor.h:
```diff
@@ -254,28 +254,56 @@ using supported_primitive_arg_types = guts::typelist::typelist<
 
   template<class Functor, bool AllowDeprecatedTypes>
-  typename guts::infer_function_traits_t<Functor>::return_type call_functor_with_args_from_stack(Functor* functor, Stack* stack) {
+  std::remove_cv_t<std::remove_reference_t<typename guts::infer_function_traits_t<Functor>::return_type>> call_functor_with_args_from_stack(Functor* functor, Stack* stack) {
```
changing to `decay_t`

smessmer committed:
Update on "Allow Tensor& in the unboxing logic"


Previously, operators taking Tensor& as arguments or returning it couldn't be c10-full because the unboxing logic didn't support it.
This adds temporary support for that. We're planning to remove this again later, but for now we need it to make those ops c10-full.
See https://docs.google.com/document/d/19thMVO10yMZA_dQRoB7H9nTPw_ldLjUADGjpvDmH0TQ for the full plan.

This PR also makes some ops c10-full that now can be.

Differential Revision: [D22989242](https://our.internmc.facebook.com/intern/diff/D22989242/)

[ghstack-poisoned]

smessmer committed:
Update on "Allow Tensor& in the unboxing logic"


Previously, operators taking Tensor& as arguments or returning it couldn't be c10-full because the unboxing logic didn't support it.
This adds temporary support for that. We're planning to remove this again later, but for now we need it to make those ops c10-full.
See https://docs.google.com/document/d/19thMVO10yMZA_dQRoB7H9nTPw_ldLjUADGjpvDmH0TQ for the full plan.

This PR also makes some ops c10-full that now can be.

Differential Revision: [D22989242](https://our.internmc.facebook.com/intern/diff/D22989242/)

[ghstack-poisoned]
